### DISCUSSION TITLE
Questions about parity in e3nn

### DISCUSSION THREAD

**Discussion by tomato1mule** [e3nn/e3nn#343] [Q&A] [answered]
Thanks for releasing this great and easy-to-use package! I have some questions about the parity of the representation.

I am currently working on a data where chirality matters, so I want a way to get rid of the unwanted parity equivariance and only maintains SE(3) equivariance. Simplest approach would be just adding up even and odd vectors of the same degree, but this seems to be a very inefficient way.

I saw the answer by Mr. Geiger to the same question in this discussion: https://github.com/e3nn/e3nn/discussions/287 

> If you declare all the irreps as even (the irreps of spherical harmonics becomes 0e+1e+2e...) your network becomes equivariant to SE(3) instead of E(3). If you are equivariant to SE(3) the two chiral shapes are two different objects related by no symmetries.

But for me it's still not obvious how using only even irreps can be a solution. For example, let's say that the input pointcloud X is passed through the network to produce a pseudovector(1e) Y=f(X). If we apply parity transform on input X, then by the parity equivariance, f(PX)=PY=Y. This is clearly an unwanted symmetry, since if the network is only SE(3) equivariant, then f(PX)=Z, where Z can be arbitrarily different w.r.t. Y.

Thank you!

**Comment by mariogeiger** [comment, marked as answer]
If you impose the irreps of the spherical harmonics to be all even, it's equivalent of saying that your input relative positions are even vectors. Therefore the whole input is even wrt P and the formula f(Px)=Py becomes trivial because Px=x and Py=y. 

Said otherwise, by using only even parities you make P having no effect. 

In this context, inverting the sign of the input position would not be an operation that is part of the symmetries of your system. 

I would also like to say that equivariance to parity is usually compatible with prediction based on chirality and I think that if you think more about it you might want parity for your problem. Remember that you can output pseudo scalars (l=0 and p=-1).

**Reply by tomato1mule** [reply]
Thank you very much for the answer!
Unfortunately, my question is still not resolved. I understand that by imposing all the irreps of SH to be even, the parity operation becomes trivial. However, isn't it restricting the expressivity of the network too much?

For example, let's say that the input is a pointcloud of my left or right hand, and the output is a scalar(0e).
Let's say that the input pointcloud of the right hand is X_r and the left hand is X_l.
I want f(X_r) = 1.478 and f(X_l) = 6.292 (The numbers are arbitrarily chosen).
However, such a network cannot be constructed with even irreps only, because PX_r = X_l and vice versa, and f(PX_r)=P(1.478)=1.478 but f(PX_r)=f(X_l)=6.292 so it's a contradiction.

Since X_r and X_l are completely different object in the context of SE(3) equivariance, so the network should be able to discern X_r and X_l, but by using even irreps only, the network cannot distinguish between them, which severely restricts the expressivity of the network.

**Reply by tomato1mule** [reply]
I think I have mistaken and you are right. 
By declaring all the irreps to be even, one can obtain every possible SO(3) equivariant pathways for the tensor product.

**Comment by blondegeek** [comment]
Hi @tomato1mule,

Parity does matter even if you use the full irreps of O(3), so will likely still want to use both even and odd irreps. Using the O(3) irreps encode parity equivariance (not invariance). However, it might change how you set up your training task!

For example, in the tetris examples, we classify the 8 3D tetris pieces but we do so with 6 scalars and 1 pseudoscalar (https://github.com/e3nn/e3nn/blob/main/examples/tetris_gate.py#L104) -- this is because two of the tetris pieces are chiral versions of each other. We only need to train on one of them to identify both and the pseudoscalar will be +1 for one of them and -1 for the other. So as you can see, you can still be sensitive to parity with the full irreps -- but you may need different irreps for your task to allow the network to show you when it sees something of the opposite parity.

Hope that helps!
Tess

**Reply by tomato1mule** [reply]
Thank you very much for your reply! My problem has been resolved.
